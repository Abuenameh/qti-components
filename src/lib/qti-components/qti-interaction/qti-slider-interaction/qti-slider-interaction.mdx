import { Canvas, Meta } from '@storybook/blocks';

import * as Stories from './qti-slider-interaction.stories';

<Meta title="qti-interactions/qti-slider-interaction" />

# qti-slider-interaction

<Canvas of={Stories.Default} />
