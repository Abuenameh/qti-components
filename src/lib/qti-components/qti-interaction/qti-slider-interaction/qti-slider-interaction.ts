import { css, html } from 'lit';
import { state, property, query } from 'lit/decorators.js';

import { Interaction } from '../internal/interaction/interaction';
import { watch } from '../../utilities/decorators/watch';

export class QtiSliderInteraction extends Interaction {
  @query('#knob')
  private _knob: HTMLElement;

  @query('#rail')
  private _rail: HTMLElement;

  @property({ type: Number }) value: number;

  @property({ type: Boolean, attribute: 'step-label' }) stepLabel = false;

  @property({ type: Boolean }) reverse = false;

  private _min: number;
  @property({ type: Number, attribute: 'lower-bound' }) set min(value: number) {
    this._min = value;
    this.value = value;
    this.style.setProperty('--min', `${this._min}`);
  }
  get min(): number {
    return this._min;
  }

  private _max: number;
  @property({ type: Number, attribute: 'upper-bound' }) set max(value: number) {
    this._max = value;
    this.style.setProperty('--max', `${this._max}`);
  }
  get max(): number {
    return this._max;
  }

  private _step: number;
  @property({ type: Number, attribute: 'step' }) set step(value: number) {
    this._step = value;
    this.style.setProperty('--step', `${this._step}`);
  }
  get step(): number {
    return this._step;
  }

  @watch('disabled', { waitUntilFirstUpdate: true })
  _handleDisabledChange = (old, disabled) => {};

  @watch('readonly', { waitUntilFirstUpdate: true })
  _handleReadonlyChange = (old, readonly) => {};

  reset() {
    // throw new Error('Method not implemented.');
  }
  validate(): boolean {
    return true;
  }

  set response(myResponse: string | string[]) {
    if (Array.isArray(myResponse)) {
      console.error('QtiSliderInteraction: response is an array, but should be a single value');
      return;
    }
    const value = parseInt(myResponse);
    if (Number.isNaN(value)) {
      console.error('QtiSliderInteraction: response is not a number');
      return;
    }
    this.value = value;
  }

  static override styles = [css``];

  override render() {
    // convert the value, which is the real slider value to a percentage for the dom.
    this.value < this.min && (this.value = this.min);
    this.value > this.max && (this.value = this.max);
    const _leftValue = ((this.value - this.min) / (this.max - this.min)) * 100;

    return html`<slot name="prompt"></slot>
      <div id="slider" part="slider">
        <div id="bounds" part="bounds">
          <div>${this._min}</div>
          <div>${this._max}</div>
        </div>
        <div id="ticks" part="ticks"></div>
        <div id="rail" part="rail" @mousedown=${this._onMouseDown} @touchstart=${this._onTouchMove}>
          <div id="knob" part="knob" style="left:${_leftValue}%"></div>
        </div>
        <div id="value" part="value">${this.value}</div>
      </div>`;
  }

  override connectedCallback() {
    super.connectedCallback();
    this.step = 1;
    this.setAttribute('tabindex', '0');
    this.setAttribute('role', 'slider');
  }

  private _onTouchMove(event) {
    const handleTouchMove = event => {
      const { x } = this.getPositionFromEvent(event);
      const diffX = x - this._rail.getBoundingClientRect().left - document.documentElement.scrollLeft;
      this.calculateValue(diffX);
      event.stopPropagation();
    };

    const handleTouchEnd = () => {
      document.removeEventListener('touchmove', handleTouchMove);
      document.removeEventListener('touchend', handleTouchEnd);

      this.saveResponse(this.value.toString());
    };

    document.addEventListener('touchmove', handleTouchMove);
    document.addEventListener('touchend', handleTouchEnd);

    const { x } = this.getPositionFromEvent(event);
    const diffX = x - this._rail.getBoundingClientRect().left - document.documentElement.scrollLeft;
    this.calculateValue(diffX);
    event.stopPropagation();
  }

  private _onMouseDown(event) {
    const handleMouseMove = (event: MouseEvent) => {
      // PK: Fixme.. this fixes
      // if the qti-slider-interaction has an absolute left position and body is scrolled a bit, take account for that
      const diffX = event.pageX - this._rail.getBoundingClientRect().left - document.documentElement.scrollLeft;

      this.calculateValue(diffX);
      event.preventDefault();
      event.stopPropagation();
    };

    const handleMouseUp = () => {
      document.removeEventListener('mousemove', handleMouseMove);
      document.removeEventListener('mouseup', handleMouseUp);

      this.saveResponse(this.value.toString());
    };

    document.addEventListener('mousemove', handleMouseMove);
    document.addEventListener('mouseup', handleMouseUp);

    // PK: Fixme.. this fixes
    // if the qti-slider-interaction has an absolute left position and body is scrolled a bit, take account for that
    const diffX = event.pageX - this._rail.getBoundingClientRect().left - document.documentElement.scrollLeft;

    this.calculateValue(diffX);
    event.preventDefault();
    event.stopPropagation();
  }

  /** calculateValue gets x position and compares this with the total width in pixels */
  private calculateValue(diffX: number) {
    const valueNow = this.min + ((this.max - this.min) * diffX) / this._rail.getBoundingClientRect().width;
    const roundedStepValue = this.min + Math.round((valueNow - this.min) / this._step) * this._step;
    this.value = roundedStepValue;
  }

  private getPositionFromEvent(e: any): {
    x: number;
    y: number;
  } {
    let _touchMove;
    if (e.type == 'touchstart' || e.type == 'touchmove' || e.type == 'touchend' || e.type == 'touchcancel') {
      const evt = typeof e.originalEvent === 'undefined' ? e : e.originalEvent;
      const touch = evt.touches[0] || evt.changedTouches[0];
      _touchMove = {
        x: touch.pageX,
        y: touch.pageY
      };
    } else if (
      e.type == 'mousedown' ||
      e.type == 'mouseup' ||
      e.type == 'mousemove' ||
      e.type == 'mouseover' ||
      e.type == 'mouseout' ||
      e.type == 'mouseenter' ||
      e.type == 'mouseleave'
    ) {
      _touchMove = {
        x: e.clientX,
        y: e.clientY
      };
    }
    return _touchMove;
  }
}

customElements.define('qti-slider-interaction', QtiSliderInteraction);
