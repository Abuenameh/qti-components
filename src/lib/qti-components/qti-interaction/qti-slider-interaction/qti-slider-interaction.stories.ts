import { html } from 'lit';
import { action } from '@storybook/addon-actions';

import './qti-slider-interaction';
import { ifDefined } from 'lit/directives/if-defined.js';

import * as htmlToImage from 'html-to-image';
import { toPng, toJpeg, toBlob, toPixelData, toSvg } from 'html-to-image';
import { ref } from 'lit/directives/ref';

export default {
  component: 'qti-slider-interaction'
};

type Interaction = {
  min: number;
  max: number;
  step: number;
  orientation: 'horizontal' | 'vertical';
  stepLabel: boolean;
  reverse: boolean;
  response: string;
  readonly: boolean;
  disabled: boolean;
};

export const Interaction = {
  render: args => {
    const { min, max, step, orientation, stepLabel, reverse, response, readonly, disabled } = args;

    const showScreenshot = () => {
      const interaction: HTMLElement = document.querySelector('qti-slider-interaction');
      const shot: HTMLImageElement = document.querySelector('.screenshot');

      htmlToImage
        .toPng(interaction, {
          width: parseInt(interaction.style.width),
          height: parseInt(interaction.style.height),
          pixelRatio: 2
        })
        .then(function (dataUrl) {
          shot.src = dataUrl;
        })
        .catch(function (error) {
          console.error('oops, something went wrong!', error);
        });
    };

    return html`
      <button @click=${() => showScreenshot()}>screenshot</button>

      <!-- style="width:300px; left:100px; position: absolute" -->
      <qti-slider-interaction
        @qti-interaction-response="${action(`on-interaction-response`)}"
        ?disabled=${disabled}
        ?readonly=${readonly}
        response=${ifDefined(response)}
        response-identifier="RESPONSE"
        lower-bound=${min ? ifDefined(min) : 0}
        upper-bound=${max ? ifDefined(max) : 10}
        step=${ifDefined(step)}
        orientation=${ifDefined(orientation)}
        ?step-label=${stepLabel}
        ?reverse=${reverse}
      >
        <qti-prompt>
          In total, what percentage of the UK population do you think were eventually classified as having no religion?
        </qti-prompt>
      </qti-slider-interaction>
      <img class="screenshot" />
    `;
  },

  args: {}
};
