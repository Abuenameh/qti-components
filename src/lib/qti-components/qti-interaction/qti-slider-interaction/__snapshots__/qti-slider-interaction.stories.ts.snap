// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/qti-components/qti-interaction/qti-slider-interaction Default smoke-test 1`] = `
<div id="root-inner">
  <qti-slider-interaction response-identifier="RESPONSE"
                          lower-bound="0"
                          upper-bound="10"
                          tabindex="0"
                          role="slider"
                          aria-valuenow="0"
                          style="--min: 0; --max: 10; --step: 1; --value-percentage: 0%;"
  >
    <qti-prompt>
      In total, what percentage of the UK population do you think were eventually classified as having no religion?
    </qti-prompt>
  </qti-slider-interaction>
</div>
`;
